File: evaluate_reverse_polish_notation.go
func evalRPN(tokens []string) int {
    stack := []int{}

    for i := 0; i < len(tokens); i++ {
        if isOperator(tokens[i]) {
            doOperation(&stack, tokens[i])
        } else {
            num, _ := strconv.Atoi(tokens[i])
            stack = append(stack, num)
        }
    }

    return stack[0]
}

func isOperator(op string) bool {
    if op == "+" || op == "-" || op == "*" || op == "/" {
        return true
    }
    return false
}

func doOperation(stack *[]int, op string) {
    num2 := (*stack)[len(*stack)-1]
    num1 := (*stack)[len(*stack)-2]
    *stack = (*stack)[:len(*stack)-2]

    res := 0
    if op == "+" {
        res = num1 + num2
    } else if op == "-" {
        res = num1 - num2
    } else if op == "*" {
        res = num1 * num2
    } else {
        res = num1 / num2
    }
    *stack = append(*stack, res)
}

// Time Complexity : O(n)
// Space Complexity : O(n)

File: evaluate_reverse_polish_notation.cpp
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<int> solver;
        int ans;
        if(tokens.size() == 1) return stoi(tokens[0]);

        for(auto token: tokens) {
            if(token == "+" || token == "-" || token == "*" || token == "/") {
                int b = solver.top();
                solver.pop();
                int a = solver.top();
                solver.pop();

                if(token == "+") ans = a+b;
                else if(token == "-") ans = a-b;
                else if(token == "*") ans = a*b;
                else ans = a/b;

                solver.push(ans);
            }
            else solver.push(stoi(token));
        }
        return ans;
    }
};

// Time Complexity : O(n)
// Space Complexity : O(n)
